### ISSUE TITLE
Paddy 2 vs Anima Toolkit

### ISSUE THREAD

**Issue by stigmates** [DWilliames/paddy2-beta#56]
Hey,

What a pain… hat a pain… how can it be? do you aim at compete with Anima autolayout?!
Heck! how can we work by now? we do need bith of the plugin. Our design system has numerous symbols using paddy and autolayout. 
We had big issues with Paddy 1 earlier this year and now, just being happy to welcome back paddy 2, it messes once again! 
We CAN'T work this way. I'm in worldwide company. We CAN't afford wasting time because of messy plugins.
Think about it.
best,

**Comment by blaskotron** [comment]
What is the issue? .6 just came out a couple hours ago which had some fixes.

**Comment by highvoltaged** [comment]
This is a free plugin someone makes in his free time.
Think about it.

**Comment by stigmates** [comment]
The issue? guess you know it. :-)
When you activate the last paddy version here is what you get:
"Unfortunately, Paddy doesn't play nice with Anima Toolkit; so you're going to have to make the difficult choice of using one over the other..."

Regarding the big evolution of paddy and the common functions with autolayout, I don't think it's preposturous to think you might try to challenge Anima et get the first place…
But if it is true, it's quite a hard way to behave with designers…

**Comment by ajilderda** [comment]
Oh wow, that is really disrespectful. The issue is you have installed a beta version and don't know what that means. It means it's likely you'll experience bugs. Maybe you shouldn't rely on beta software for your worldwide company. 

If you find yourself wasting too much time messing around with messy plugins let me give you a hint: invest that time in creating your own. 

I'd say this issue can be closed.

**Comment by ajilderda** [comment]
I think my point is still valid. Don't rely that much on beta plugins for your workflow (cause the chance is big you'll run into issues). 

A beta version is helpful for developers so that others can help find and file issues. Your 'issue' sounds more like a complaint, which is invalid since this plugin is free (!) and is still in beta. 

If you describe the issues you're having, David might be able to fix it and release a stable version sooner.

**Comment by aparajita** [comment]
@stigmates Are you willing to give hundreds of hours of your time for free to your clients? If not, then why are you expecting perfection from someone who is giving hundreds of hours of his valuable time to you?

Paddy 2 is in beta, and he clearly states that it might have bugs. If you want better quality in Paddy 2, then consider donating generously to its development. If you find it isn't stable enough for your incredibly important worldwide design company, then don't use it. No one is forcing you to. You didn't pay anything for it. Is it fair for you to attack him like this?

David is **not** trying to compete with Anima, he's just doing what interests him. What he said is exactly right: "Unfortunately, Paddy doesn't play nice with Anima Toolkit." Of course they don't play nice together, they are both trying to do the same thing to layers, which is bound to cause problems. So what he said is a simple statement of fact, not of competition. Maybe you think that way, but David doesn't. He's just having fun... or perhaps he was, until now.

**Comment by godyj** [comment]
@stigmates I support what @aparajita said above. And to add more to that …
This is a **labor of love from David**. I do not believe you paid him anything for his plugin. BE POLITE and ask for help – DO NOT DEMAND. Even if you paid for something, BE POLITE. You're not going to get far by being rude. This is in beta and we are all working together with David to help iron out the bugs. If that's too much work for you, stop using it, and stop being rude!

**Comment by DWilliames** [comment]
Thanks all for the responses and love in response. And thank you for taking your time to clarify for @stigmates.

I still think I might be able to shed a bit of light here, and offer a bit more clarity and transparency to everyone regarding 'Anima vs Paddy 2'. Bare with me as I try to provide more information on the matter...

---

### Messy/buggy plugin?
@stigmates Firstly in response to; ***'We CAN'T work this way. I'm in worldwide company. We CAN't afford wasting time because of messy plugins.'*** 

Whilst I do understand your frustration, I have clearly stated that this is a Beta plugin and ***"Warning: Please use the Beta version at your own risk. I don't want be responsible for possibly 'ruining' your very important Sketch document."*** So whilst I appreciate you giving Paddy 2 a go; I would not advise integrating it with your projects within your 'worldwide company'... yet. I am in the process of trying to iron out these bugs; hence why it is in **beta.**

### Competing with Anima Toolkit?
Whilst the question wasn't asked in the most polite way; it is a very valid point. Let me answer you...

#### Why doesn't Paddy 2 work with Anima? 
In order to integrate with Sketch the way Paddy 2 does, it uses some methods of 'swizzling', which allows me to hook in some code to specific actions that Sketch uses internally. I do this in a way that shouldn't affect other plugins...

Anima hooks into internal actions too via a framework called Aspects. Using this framework makes it much easier for them, and gives a higher level of control — which makes perfect sense as to why the'd use it. Unfortunately, since they use Aspects, it doesn't allow other plugins to be able to swizzle the same actions they do... meaning Paddy 2 wouldn't be able to be automated or function in some cases when working around this constraint.

I got in contact with the developers of Anima to see if it was at all possible for them to move off the Aspects framework, or if they knew a way that I could work with it. Which was a massive ask — and completely understandably they responding saying that after investigation, they aren't able to move off the framework.

I have spent numerous hours & hours trying to work around this limitation to no avail. :( 

#### What's my approach?
Since it is completely unavoidable that the 2 plugins cannot work together, there's a couple of options I have here...
1. I could remove all the functionality that integrates deeply into Sketch to work around the constraints, creating a much less valuable experience, if not unusable.
2. I could just give up... given your philosophy @stigmates that you wouldn't release a plugin like this with such a huge limitation to professionals — then that would be the only option.
3. Try and build out as much functionality as I can to fill the gap that not using Anima may have on users. I know a lot of people use Anima purely for Stack Groups; and their ability to stack content even within a symbol instance... so I am going to make sure that functionality is built out in Paddy 2.

#### What's next?
There's a couple of things I still aim to do to make the transition from Anima to Paddy 2 for users easier, if they make that decision.
* Allow stack groups to be rendered dynamically within a symbol instance — whilst very tricky, this is a key differentiator for Anima's stack groups right now. But rest assured, that is coming for Paddy 2.
* Create an action to convert existing Anima Stack groups to Paddy Stack groups; which will hopefully make that transition easier

### Conclusion
The fact that I have not allowed them to work together is not a choice! I spent ages trying to make it happen... it's simply unavoidable.

**Using Paddy 2 is optional.** Whilst not ideal, users are going to have to make the difficult choice of using one plugin over the other. Paddy 2 will not in any way be able to replace Anima Toolkit overall; but I am hoping that it will live up to the task of replacing Stack Groups.

It doesn't look like this is a bug I will be able to work around any time soon... if anyone knows a way, please let me know.

Thanks all for your understanding.

---

### TL;DR
* Using Paddy 2 is optional
* This is not a choice; it is a technical constraint that it appears I won't be able to work around
* I've tried and tried to work fix it and reached out to Anima developers – but no fix seems available
* I will do all I can to make the transition for people who rely on Anima Stack Groups easier

**Comment by stigmates** [comment]
@DWilliames Thank you for your response and explanation. I understand the choices you have to make. Working with a everyday evolving tool (Sketch) is quite a new way to do for designers. But not only; it's concerning techs and industry in a wider way. It's cool but sometimes uncomfortable. And releasing unfinished device or software (MVPs, but sometimes not so viable…) I think it is a real issue of contemporary way of producing.
Sometimes it's hard to rely on constant moving tools to build a lasting solution (design system or whatever). I already explained why I chose to go with paddy 2 whilst it's beta. I didn't and wouldn't build the whole library with it of course. But key symbols have been designed with previous versions… and it was a time saviour (sometimes not…).
I was disappointed because paddy and Anima were complementary and my team and I were linked to the both plugins. We had numerous troubles to deal with (craft issues, zeplin exports, sketch updates…) and now we have to chosse between the bothn plugins… 'la goutte d'eau qui fait déborder la vase' (in french).
Typing instead of talking often turns terms harsher than they really are, especially when you write in a foreign language. Maybe several users had the same thoughts and reaction, but they didn't tell.
Let's close this debate and please accept my apologies if I hurt you.

**Comment by DWilliames** [comment]
Thanks for understanding. 

It’s tough with so much fluidity with design tools and plugins right now. Stability and compatibility is very sought after, I understand that. 

I definitely don’t see this as a debate; but more of an opportunity for me to have open and honest dialogue. 

I feel that perhaps ‘Beta’ may have been too generous a term for ‘Paddy 2’ — in reality it’s more of an ‘Alpha release’ or simply a ‘tech preview’. For me, I’ve always struggled to complete projects as I have a constant urge for it to be ‘perfect’ before showing others... so this was a big step for me to put Paddy 2 out there in the open as I work through it, in its infancy. And letting everyone come on the journey with me. Perhaps I should have framed the context a bit better so that others are also not as frustrated with a ‘buggy’ or ‘unstable’ plugin. 

I can without a doubt say that the support from everyone has been incredible and definitely helped me push forward on this project. I gave it a break for a couple of weeks; but now I’m back on it, pushing full steam ahead. 🙌

Thanks everyone for your patience and understanding as I try make this thing better. 👌🤗❤️

**Comment by vic-topcoder** [comment]
I think this is one of the most important discussions as the future of the Paddy 2 is at stake as a successful plugin. There have been numerous projects (including the native resizing behavior of Sketch) that address flexible layout and layout constraints. To this day none of the solutions provides a comprehensive 100% cover of all the layout cases we face. It is impossible to build a button that autoresizes just like that, which is extremely frustrating for UI work.

I don't really need the additional functionality bundled with Anima's auto layout (HTML export, timeline) as I really would like to have a complex layout constraints system that does only that, and works well with Sketch. I hope that at some point Bohemian would open up the system to support auto layout on a new level that would make working with higher abstraction elements reality (I think of them as components - buttons, input fields, repeater grids, etc.). This is what we face as UI designers that deal with multiple screen resolutions and devices.

So the question for Paddy 2 is really - where to? Should Paddy 2 become the ultimate auto-layout engine, or just a complimentary plugin? I would firmly prefer to have a single solution! Hope @DWilliames really succeeds in this very hard undertaking!

**Comment by pietrogregorini** [comment]
Well, guys, let's talk straight: this kind of functionality should be integrated natively by Sketch because nowadays is kinda mandatory for people working on UI like us. Bohemian should seriously think about this... Or just hiring @DWilliames 😉 

By the way, @DWilliames thanks for all the hard work ❤️Can't wait to use Paddy 2!
